stock_storage_type_buffer/models/stock_location.py

```diff
@@ -0,0 +1,54 @@
+# Copyright 2020 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+
+import logging
+
+from odoo import models
+
+_logger = logging.getLogger(__name__)
+
+
+class StockLocation(models.Model):
+    _inherit = "stock.location"
+
+    def _select_final_valid_putaway_locations(self, limit=None):
+        """Return the valid locations using the provided limit
+
+        ``self`` contains locations already ordered and contains
+        only valid locations.
+        This method can be used as a hook to add or remove valid
+        locations based on other properties. Pay attention to
+        keep the order.
+        """
+        if self:  # short-circuit computation when already 0 valid
+            self = self._filter_locations_blocked_by_buffer()
+        return super()._select_final_valid_putaway_locations(limit=limit)
+
+    def _filter_locations_blocked_by_buffer(self):
+        valid_location_ids = set(self.ids)
+        for storage_buffer in self.env["stock.location.storage.buffer"].search([]):
```

Comment:
You compute all the buffers availability even if the valid locations are not guarded by any buffer. I would have done it the other way around by checking each valid location if it is guarded by a buffer and filter it out if the buffer is not empty.
The pro:
- You compute only the required buffers. For each buffer location_is_empty should only be computed once if kept in cache.
- If you define the buffer on the location, you can read_group it at once to not read each location one by one which would be costly
- If there is a limit, you can stop the iteration on the locations when the limit is reached

Reply:
We have to measure the cost of the buffer availability, we don't have a lot of them and it only queries indexed relations.

Reply:
The cost of `if storage_buffer.buffers_have_capacity():` is really low (between 0.01 and 0.06 per buffer on my test database after stress test), so for now I don't think it's worth changing this.